### ISSUE TITLE
Can't add new Unity installations

### ISSUE THREAD

**Issue by appkaiju** [flathub/com.unity.UnityHub#47]
Before 2.4.1 I used this Flatpak version to install multiple Unity Editors and launch the Unity Hub with no problems whatsoever.
But with 2.4.1 I can't add new Unity Editors.
I have more than 500 GB of free disk space but it complains about not having enough free space.
![Screenshot from 2020-10-15 18-20-59](https://user-images.githubusercontent.com/71205864/96187432-5fbdc680-0f13-11eb-9367-5822654cc155.png)
![Screenshot from 2020-10-15 18-21-03](https://user-images.githubusercontent.com/71205864/96187442-63514d80-0f13-11eb-8a8c-b89be3e4970b.png)
What should I do? Please inform me if you need specific data about my specs.
```
OS: Linux Mint 20 x86_64 
Kernel: 5.4.0-51-generic 
Packages: 2306 (dpkg), 47 (flatpak) 
DE: Cinnamon 
WM: Mutter (Muffin) 
CPU: Intel i7-7500U (4) @ 3.500GHz 
GPU: NVIDIA GeForce 940MX 
Memory: 2660MiB / 15899MiB
``` 
Edit: corrected typo.

**Comment by subjectdenied** [comment]
same here, it seems it is again about disk space

**Comment by refi64** [comment]
I believe I've fixed this locally, but for some reason I just *cannot* push
to GitHub right now. Will try again tomorrow...

**Comment by refi64** [comment]
hahaha looks like it was just GNOME keyring freaking out, just pushed and
will update here once the builds are published.

**Comment by refi64** [comment]
@appzillamonster @subjectdenied please try the latest version now

**Comment by appzillamonster** [comment]
Thank you for your swift response, @refi64. Regrettably my Unity Hub version didn't update, according to `flatpak list` it stays at version 2.4.1. I tried running `flatpak update` but flatpak says that there are no updates avaible. Looking at Flathub https://flathub.org/apps/details/com.unity.UnityHub I see that the distributable was updated today, October 17, but extrangely the version is the same. Please tell me if I'm missing something.

**Comment by refi64** [comment]
@appzillamonster ah let me clarify: the Unity Hub version itself is unchanged, but it got an update from me removing an old workaround for a bug that seems to have since been fixed in Hub itself. I.e. the app itself doesn't have a new version, but the packaging was updated.

**Comment by appzillamonster** [comment]
> @appzillamonster the app itself doesn't have a new version, but the packaging was updated.
It makes sense, I runned `flatpak history` and I see it flatpak pulled a new package yesterday. Unfortunately, the issue persists. I don't see any difference in behaviour. Please inform me if you need further data from me.

**Comment by subjectdenied** [comment]
@refi64 can confirm that the issue persists for me too

**Comment by subjectdenied** [comment]
@refi64 if it helps, i am using fedora 33 workstation

**Comment by crolar2015** [comment]
Same happens on PopOS.

**Comment by Aurrix** [comment]
Guys can somebody look into an issue with Pop OS?

**Comment by jasonboukheir** [comment]
Same on PopOS

**Comment by KevinNThomas** [comment]
Same thing happens to me on Gentoo

**Comment by CDAGaming** [comment]
Same issue with PopOS for the last few months, both on flatpak and the popos store

**Comment by mucahitkmlay** [comment]
> Before 2.4.1 I used this Flatpak version to install multiple Unity Editors and launch the Unity Hub with no problems whatsoever.
> But with 2.4.1 I can't add new Unity Editors.
> I have more than 500 GB of free disk space but it complains about not having enough free space.
> ![Screenshot from 2020-10-15 18-20-59](https://user-images.githubusercontent.com/71205864/96187432-5fbdc680-0f13-11eb-9367-5822654cc155.png)
> ![Screenshot from 2020-10-15 18-21-03](https://user-images.githubusercontent.com/71205864/96187442-63514d80-0f13-11eb-8a8c-b89be3e4970b.png)
> What should I do? Please inform me if you need specific data about my specs.
> 
> ```
> OS: Linux Mint 20 x86_64 
> Kernel: 5.4.0-51-generic 
> Packages: 2306 (dpkg), 47 (flatpak) 
> DE: Cinnamon 
> WM: Mutter (Muffin) 
> CPU: Intel i7-7500U (4) @ 3.500GHz 
> GPU: NVIDIA GeForce 940MX 
> Memory: 2660MiB / 15899MiB
> ```
> 
> Edit: corrected typo.

i fix this; Do not install unity on flatpak, you can install here https://store.unity.com/download-nuo and then use this command 
chmod +x UnityHub.AppImage and then ./sh UnityHub.AppImage

**Comment by williamd1k0** [comment]
>i fix this; Do not install unity on flatpak, you can install here https://store.unity.com/download-nuo and then use this command
chmod +x UnityHub.AppImage and then ./sh UnityHub.AppImage

You haven't fixed it. Official UnityHub has very poor DE integration, it does not support xdg-open, for example.

**Comment by CDAGaming** [comment]
> >i fix this; Do not install unity on flatpak, you can install here https://store.unity.com/download-nuo and then use this command
> chmod +x UnityHub.AppImage and then ./sh UnityHub.AppImage
> 
> You haven't fixed it. Official UnityHub has very poor DE integration, it does not support xdg-open, for example.

Can second this, both editions have this issue for me.

**Comment by OPuffer** [comment]
I am having this issue too. Best I can tell it has something to do with how UnityHub wants to download all things in our tmp folder before moving the files elsewhere. On pop OS (And maybe other distros too, i'm new to linux) the temp folder is too small, so unity complains about lack of space. I'm not sure how to get the unityHub application to use a different folder as its initial download location, but maybe one of you smarter folks can figure it out.

**Comment by DrEMartstud** [comment]
Just install it from the terminal like this good sir told you
chmod +x UnityHub.AppImage
then
./UnityHub.AppImage

> command
> `chmod +x UnityHub.AppImage` and then `./sh UnityHub.AppImage`

**Comment by CDAGaming** [comment]
> Just install it from the terminal like this good sir told you
> chmod +x UnityHub.AppImage
> then
> ./UnityHub.AppImage
> 
> > command
> > `chmod +x UnityHub.AppImage` and then `./sh UnityHub.AppImage`
> 
>  This basically fixed my issue
> 

This has been tried over and over again. It fixes it by displaying the correct size, but will still fail with an error due to the aame backend tmp issue.

I dont think we should keep running arround in circles suggesting this or thisll never get a full solution.

**Comment by OPuffer** [comment]
I managed to download a unity version by using the AppImage version of the launcher from the unity website. From there, this version works to launch unity properly. 

First I created a tmp folder in my home directory, then I navigated to the appImage section and ran

TEMP=~/tmp ./UnityHub.AppImage

when I wanted to install a version from the online archive I ran 
TEMP=~/tmp ./UnityHub.AppImage <The the actual url of the green unityHub download button link>

so for me installing 2019.3.14 that was 
TEMP=~/tmp ./UnityHub.AppImage unityhub://2019.3.14f1/2b330bf6d2d8

Then I was able to exit that and use this version to launch the unity version. This workaround worked for me, but since I am still a Linux Noob, I am unsure what the repercussions of this workaround are. Would It be possible to bake something like this into this install, or is that going to cause a whole host of other problems?

If anything hopefully this helps someone get it working for now.

**Comment by djotaku** [comment]
Same issue for me on flathub version of Unity.

**Comment by williamd1k0** [comment]
> @OPuffer  I am having this issue too. Best I can tell it has something to do with how UnityHub wants to download all things in our tmp folder before moving the files elsewhere. On pop OS (And maybe other distros too, i'm new to linux) the temp folder is too small, so unity complains about lack of space. I'm not sure how to get the unityHub application to use a different folder as its initial download location, but maybe one of you smarter folks can figure it out.

Ah, I think I understand now what the problem might be.
The /tmp directory on linux is usually mounted using tmpfs, so it will use your memory (ram) instead of storage (hd/ssd).

**Comment by refi64** [comment]
We *should* have TMPDIR underneath `$XDG_CACHE_HOME` though 🤔

**Comment by davidalmarinho** [comment]
Same happening here in:
NAME="Pop!_OS"
VERSION="20.10"

**Comment by AhmetEminAyar** [comment]
I have the same problem on Pop Os 20.04

**Comment by robigan** [comment]
Same issue here. My research so far has come up that by default debian limits the size of /tmp. I am not exactly a Linux master and for some reason /tmp is not loaded as tmpfs so doing sudo mount remount /tmp thingy doesn't work as it's not mounted nor do I know how to fix that. As of writing am in online classes and modifying /etc/fstab to change the size of /tmp and then restart is both undesired bc it's not every day I have to upgrade unity and don't wanna touch the default size. I tried to run TEMP=/home/robigan/Unity/Tmp flatpak run com.unity.UnityHub but that doesn't seem to do jack s***. HOLY SHIT, at this very instant I did what OPuffer did and this let me install. Ngl I was kinda surprised

**Comment by springbov** [comment]
Okay, so I have a hunch as to why it's not working, but I don't know 100% since I'm not super familiar with how flatpak works.

It looks like there is a wrapper script that actually starts unityhub. No matter what you do, [it will override the TMPDIR](https://github.com/flathub/com.unity.UnityHub/blob/06866de409c051773939518b68aa0db92250eafd/unityhub.py#L74) to `f'{env["XDG_RUNTIME_DIR"]}/app/{env["FLATPAK_ID"]}'` every time you run it.

I tried testing this by running a shell in the context of that flatpak by runnting

`flatpak run --command=sh com.unity.UnityHub`

and then running the command inside the container

`UNITY_DATADIR=$XDG_DATA_HOME TEMPDIR=$XDG_CACHE_HOME/tmp zypak-wrapper /app/extra/unityhub-bin`

with this I am now downloading the current version of unity. (yay, woo).

With this I think the fix is to just change [line 74 in unityhub.py](https://github.com/flathub/com.unity.UnityHub/blob/06866de409c051773939518b68aa0db92250eafd/unityhub.py#L74)
from:
```Python
    env['TMPDIR'] = f'{env["XDG_RUNTIME_DIR"]}/app/{env["FLATPAK_ID"]}'
```

to
```Python
    env['TMPDIR'] = f'{env["XDG_CACHE_HOME"]}/tmp'
```

I don't feel comfortable changing this myself just yet since I haven't read what the rest of the script is doing and I don't know flatpak well enough to know if this will actually fix it for all environments. At some future I might try and figure this out but I think it would be better for someone more familiar with python, unity and flatpak to do it.

**Comment by springbov** [comment]
I just went ahead and made a PR anyways. I built the flatpak on my machine, installed it, and it seems to work. If someone has corrections or improvements on the PR let me know.

**Comment by puttj** [comment]
A workaround I'm using temporarily on my Fedora 33 machine is to edit the tmpfs size used for eg `/run/user/1000`.
Set the `RuntimeDirectorySize` in `/etc/systemd/logind.conf` to eg `10G` (assuming you have available ram). I set it back again after installing the editor.

[How do I increase size of per-user tmpfs partition /run/user/$(id)? - Ask Ubuntu](https://askubuntu.com/questions/855740/how-do-i-increase-size-of-per-user-tmpfs-partition-run-user-id)

**Comment by refi64** [comment]
Fixed by #51.
